modules/liquidity-pools/src/lib.rs

```diff
@@ -0,0 +1,308 @@
+		currency_id: T::CurrencyId,
+	) -> result::Result<(), Error> {
+		ensure!(Self::is_owner(pool_id, &who), Error::NoPermission);
+		ensure!(T::PoolManager::can_remove(pool_id), Error::CannotRemovePool);
+
+		let pool = match <LiquidityPoolOptions<T>>::get(&pool_id, &currency_id) {
+			Some(pool) => pool,
+			None => return Err(Error::PoolNotFound),
+		};
+
+		T::MultiCurrency::deposit(currency_id, &who, pool.balance).map_err(|e| e.into())?;
+		<LiquidityPoolOptions<T>>::remove(&pool_id, &currency_id);
+
+		// remove owner reference
+		<Owners<T>>::remove(&pool_id);
+
+		Self::deposit_event(RawEvent::LiquidityPoolRemoved(who, pool_id, currency_id));
+		Ok(())
+	}
+
+	fn _deposit_liquidity(
+		who: T::AccountId,
+		pool_id: T::LiquidityPoolId,
+		currency_id: T::CurrencyId,
+		amount: T::Balance,
+	) -> result::Result<(), Error> {
+		let mut pool = match <LiquidityPoolOptions<T>>::get(&pool_id, &currency_id) {
+			Some(pool) => pool,
+			None => return Err(Error::PoolNotFound),
+		};
+
+		match pool.balance.checked_add(&amount) {
+			Some(new_balance) => {
+				// withdraw account
+				T::MultiCurrency::withdraw(currency_id, &who, amount).map_err(|e| e.into())?;
+
+				// update pool balance
+				pool.balance = new_balance;
+				<LiquidityPoolOptions<T>>::insert(&pool_id, &currency_id, pool);
+
+				Self::deposit_event(RawEvent::DepositLiquidity(who, pool_id, currency_id, amount));
+				Ok(())
+			}
+			None => Err(Error::DepositFailed),
+		}
+	}
+
+	fn _withdraw_liquidity(
+		who: T::AccountId,
+		pool_id: T::LiquidityPoolId,
+		currency_id: T::CurrencyId,
+		amount: T::Balance,
+	) -> result::Result<(), Error> {
+		ensure!(Self::is_owner(pool_id, &who), Error::NoPermission);
+
+		let mut pool = match <LiquidityPoolOptions<T>>::get(&pool_id, &currency_id) {
+			Some(pool) => pool,
+			None => return Err(Error::PoolNotFound),
+		};
+
+		match pool.balance.checked_sub(&amount) {
+			Some(new_balance) => {
+				// check minimum balance
+				if new_balance < T::ExistentialDeposit::get() {
+					return Err(Error::WithdrawFailed);
+				}
+
+				// deposit amount to account
+				T::MultiCurrency::deposit(currency_id, &who, amount).map_err(|e| e.into())?;
+
+				// update pool balance
+				pool.balance = new_balance;
+				<LiquidityPoolOptions<T>>::insert(&pool_id, &currency_id, pool);
+
+				Self::deposit_event(RawEvent::WithdrawLiquidity(who, pool_id, currency_id, amount));
+				Ok(())
+			}
+			None => Err(Error::WithdrawFailed),
+		}
+	}
+
+	fn _set_spread(
+		who: T::AccountId,
+		pool_id: T::LiquidityPoolId,
+		currency_id: T::CurrencyId,
+		ask: Permill,
+		bid: Permill,
+	) -> result::Result<(), Error> {
+		ensure!(Self::is_owner(pool_id, &who), Error::NoPermission);
+
+		let mut pool = match <LiquidityPoolOptions<T>>::get(&pool_id, &currency_id) {
+			Some(pool) => pool,
+			None => return Err(Error::PoolNotFound),
+		};
+
+		pool.bid_spread = bid;
+		pool.ask_spread = ask;
+		<LiquidityPoolOptions<T>>::insert(&pool_id, &currency_id, pool);
```

Comment:
`.mutate()` function would be more idiomatic than pulling the data out, modify and insert it again.

Reply:
@shaopengw The value is Option so it will make things more complicated. I've done some updates though